packages/sql-events/index.js

```diff
@@ -0,0 +1,140 @@
+'use strict'
+
+const MQEmitter = require('mqemitter')
+const fp = require('fastify-plugin')
+const { PassThrough } = require('stream')
+const MQEmitterRedis = require('mqemitter-redis')
+const { promisify } = require('util')
+
+async function fastifySqlEvents (app, opts) {
+  setupEmitter({ ...opts, mapper: app.platformatic, log: app.log })
+  const mq = app.platformatic.mq
+  app.addHook('onClose', () => promisify(mq.close.bind(mq)))
+}
+
+function setupEmitter ({ log, mq, mapper, connectionString }) {
+  if (connectionString) {
+    mq = MQEmitterRedis({ connectionString })
+  } else if (!mq) {
+    mq = MQEmitter()
+  }
+
+  for (const entityName of Object.keys(mapper.entities)) {
+    const entity = mapper.entities[entityName]
+    const { primaryKey } = entity
+    mapper.addEntityHooks(entityName, {
+      async save (original, data) {
+        const ctx = data.ctx
+        /* istanbul ignore next */
+        const _log = ctx?.reply?.request?.log || log
+        const res = await original(data)
+        const topic = await entity.getPublishTopic({ action: 'save', data: res, ctx })
+        if (topic) {
+          const payload = {
+            [primaryKey]: res[primaryKey]
+          }
+          _log.trace({ topic, payload }, 'publishing event')
+          await new Promise((resolve) => {
+            mq.emit({
+              topic,
+              payload
+            }, resolve)
+          })
+        }
+        return res
+      },
+
+      delete: multiField('delete'),
+      insert: multiField('save')
+    })
+
+    function multiField (action) {
+      return async function (original, data) {
+        const ctx = data.ctx
+        /* istanbul ignore next */
+        const _log = ctx?.reply?.request?.log || log
+        const fields = new Set(data.fields)
+        const res = await original({ ...data, fields: undefined })
+
+        await Promise.all(res.map(async (payload) => {
+          const topic = await entity.getPublishTopic({ action, data: payload, ctx })
```

Comment:
So you call `getPublishTopic` for each entity to check that there is a `primaryKey`? I don't understand why we can't just get publish url once at the beginning, since we know what entity it is.

Reply:
We need the id of each saved entity. In case of insert we don't have it.